python/pyspark/sql/dataframe.py

```diff
@@ -1721,7 +1721,8 @@ def toPandas(self):
         1    5    Bob
         """
         import pandas as pd
-        return pd.DataFrame.from_records(self.collect(), columns=self.columns)
+        dtype = {field.name: _to_numpy_type(field.dataType) for field in self.schema}
+        return pd.DataFrame.from_records(self.collect(), columns=self.columns).astype(dtype)
```

Comment:
converting rows to column format is also costly I think, so if users wanna performance here, they should enable the arrow optimization :)